### PR TITLE
Create base model class #159

### PR DESCRIPTION
**Pull request by maxmcd**

### CODE REVIEW

**Comment by maxmcd** [comment]
@kuanb 
There are quite a few things going on here that require sanity checks. This PR introduces BaseModel which should hopefully reduce a lot of the work in creating individual models. It also introduces slight changes to the test structure to allow for easier filtering.

## BaseModel

With the new BaseModel we now create models like this:
```javascript
class TestModel extends BaseModel {
  constructor(data) {
    super({
      data: data,
      columns: ["cmid","org"],
    })
  }
}
TestModel.primaryId = "cmid"
TestModel.tableName = "cms"
```

With that extension that model now has the following capability:
```javascript
// find by primary id
TestModel.findByID(id).then((testModelInstance) => {
  testModelInstance.cmid //outputs ids
  testModelInstance._info.columns //provides the columns passed in the constructor
})
// find by email 
TestModel.findByEmail(email)
```

Every model will get these, maybe they don't all need findByEmail, so we might remove that.

The model also has a few internal methods that we should continue extending.  `TestModel._getSingleResponse` handles checking if anything was returned from the db and surfaces an error if it's invalid in any way. No more repeated code to check the database responses for single entities. 

## Testing

I added `tests/app/model.js` and `tests/app/userModel.js`. You can run them like so:
``` bash
$ CCENV=testing mocha test/app/model.js test/app/_init.js
# or 
$ CCENV=testing mocha test/app/userModel.js test/app/_init.js
```

You can remove test/app/_init.js if you don't need to reload seed data. 

----

Is this new model ok? Do you think this will save a lot of time over the current process? Can it be improved? Do the tests run on your end? Any other questions/concerns?

File: models/base.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        throw new Error("This model needs a primaryId!")
+      }    
+  }
+
+  static _getSingleResponse(objects, fulfill, reject) {
+    if (!objects || objects.length === 0) {
+      reject(new Error("nothing returned from db"))
+    } else {
+      // return class instance
+      let instance = new this(objects[0])
```
Comment (maxmcd) [inline models/base.js]: Everything returned from the DB is an instance of the class, when we're good with this BaseModel we should update all DB calls to do the same

File: models/classes/users.js [hunk 2]
```diff
@@ -157,4 +148,7 @@ class Users {
   
 }
 
+Users.primaryId = "cmid"
+Users.tableName = "cms"
```
Comment (maxmcd) [inline models/classes/users.js]: This is a little weird, shoving the primaryId and tableName at the bottom of the file. Not sure how to improve this as static values can only be added after the class is instantiated.

File: server/app.js [hunk 3]
```diff
@@ -63,23 +63,25 @@ app.use(passport.session());
...
-    let statusCode = res.statusCode;
-    let contentLength = res.header()._headers['content-length'] || 0;
-    let userAgent = req.headers['user-agent'];
-    console.log(
-      `${ip} -- [${timestamp}] ` +
-      `${method} ${path} ${statusCode} `.magenta +
-      `${contentLength} ${milliseconds}ms `.cyan +
-      `"${userAgent}"\n`
-    );
-  });
```
Comment (maxmcd) [inline server/app.js]: Removing http logs from the testing env so that we get cleaner test logs

File: models/base.js (hunk 1 above)
Comment (kuanb) [inline models/base.js]: Do we need to validate what the `info` object looks like? Is there a comment somewhere that specifies exactly what it is that we need in `info`? I am assuming info is everything associated with the "sub" class that we would be instantiating... I assume this is very Rails-esque... But is there complexity in knowing/documenting `info` arg?

Comment (maxmcd) [inline models/base.js]: `info` is undocumented and very important. Every new model has to pass a correctly formatted `info` object to the BaseModel constructor. Would argue this isn't very Rails-esque and is kind of a hack to get all the values we need for the model. Will discuss on the call as I would be open to an alternate way to handle this.

Comment (kuanb) [inline models/base.js]: Sounds good

**Comment by maxmcd** [comment]
@kuanb Pushed up those changes from yesterday, let me know if the tests on this branch run successfully for you

**Comment by kuanb** [comment]
Damn. I'm slow here, sorry. I should have pulled this down and looked at it rather than review it from GH.

**Comment by kuanb** [comment]
Error when trying to get running locally:

```
Unhandled rejection Error: The migration directory is corrupt, the following files are missing: 20160810132045_groups.js, 20160812092104_color_tags.js, 20160812093249_color_tags_link.js, 20160812151717_color_tags_link_to_table_from_clients.js, 20160816130135_groups.js, 20160816134051_groups_last_updated.js, 20160818083638_alerts_feed.js, 20160818101431_departments.js, 20160818102222_organization_owners.js, 20160818102448_user_classes.js, 20160823124157_department_foreign_key.js, 20160823183957_phone_num_big_integer.js
...

```

**Comment by kuanb** [comment]
Working still on trying to pull this down and run it... There were a number of merge conflicts, that seem to have been manageable to resolve (I think I have resolved them safely, correctly). 

An issue remains at the moment where I have that ****ing Pool timeout from Knex. I am looking into that. Also, I get this error: ```Production env. New Relic running.
{"v":0,"level":50,"name":"newrelic_bootstrap","ho...``` Which seems to have to do with New Relic getting started in dev mode, so I think we need to make sure that New Relic does not get started (I think this is happening as a result of a bunch of the reshuffling).

**Comment by kuanb** [comment]
```

[nodemon] 1.9.1
[nodemon] to restart at any time, enter `rs`
[nodemon] watching: *.*
[nodemon] starting `node server/server.js`
Testing environment - no monitoring.
Credentials have been modified with test environment values.
Database being used: clientcomm-staging
Knex configuration environment: development
Listening on port 4000.
Knex:Error Pool2 - error: password authentication failed for user "kuan"
Knex:Error Pool2 - error: password authentication failed for user "kuan"

```

**Comment by kuanb** [comment]
Damn, still getting this Pool issue... Did we ever determine *why* this happens?

**Comment by kuanb** [comment]
Also, when I run `npm start`, a ton of garbage seems to get dumped into the `console` about god knows what... I think you had this same issue right? Strange that it happens on this PR/branch but not on the branch that I have been working on... What's different?

**Comment by kuanb** [comment]
gd figured it out, issue was I needed to run `npm install --save pg@4.5.5`

source: http://stackoverflow.com/questions/38111578/knexerror-pool2-error-password-authentication-failed-for-user

**Comment by kuanb** [comment]
Hey @maxmcd I have 2 commits I would like to include in this PR that address the issues in the above comments:

```
commit c4cd0589046c9e6c3ac726f0b6927b505e2c2fc8
Author: Kuan Butts <kuanbutts@yahoo.com>
Date:   Tue Sep 13 19:51:57 2016 -0700

    #159 mods to server that used to rely on TESTENV to change to CCENV and also addition to package json for node 6.x.x issue with connecting to pg and havign that pool time out issue

commit f5cab8f65bda4032c466cc740c1e9998d307ae30
Merge: 850f35e 7267394
Author: Kuan Butts <kuanbutts@yahoo.com>
Date:   Tue Sep 13 19:09:13 2016 -0700

    merge of issues in git pull from branch remote head from max's pr
```

Should I push?

**Pull request merged**